### PR TITLE
Track Replica hub IDs (#5360)

### PR DESCRIPTION
**Pull request by nadove-ucsc**
Connected issues: #5360

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via ZenHub
- [x] PR description links to connected issues
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] PR and connected issue are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] PR has checklist items for upgrading instructions <sub>or PR is not labeled `upgrade`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `hammerbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Title of merge commit starts with title from this PR
- [x] Added PR reference to merge commit title
- [x] Added commit title tags to merge commit title
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>

### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Build passes on GitLab `anvilprod`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvilprod`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`
- [x] Deleted PR branch from GitLab `anvilprod`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilprod` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvilprod` deployment <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by nadove-ucsc** [comment]
For WIP review, please look at the refactoring in the 1st commit. The intent is to facilitate the addition of scripted updates in the 2nd commit. I'd like feedback on the overall approach and the division of labor between the methods in the `Document` class.

File: src/azul/indexer/document.py [hunk 1]
```diff
@@ -1288,38 +1295,28 @@ def to_index(self,
                         self.to_json()
                 }
             ),
-            '_id' if bulk else 'id': self.coordinates.document_id
+            '_id' if bulk else 'id': self.coordinates.document_id,
+            **(
+                {
+                    '_op_type': self.op_type.name
+                }
+                if bulk else
```
Comment (hannes-ucsc) [inline src/azul/indexer/document.py]: Probably easier to just `if …: result['_op_type'] = ` afterwards

File: src/azul/indexer/index_service.py [hunk 2]
```diff
@@ -889,7 +890,7 @@ def _write_bulk(self, documents: Iterable[Document]):
                                   max_chunk_bytes=config.max_chunk_size)
         for success, info in response:
             op_type, info = one(info.items())
-            assert op_type in ('index', 'create', 'delete')
+            assert op_type in OpType.__members__, op_type
```
Comment (hannes-ucsc) [inline src/azul/indexer/index_service.py]: Isn't there a non-dunder way to do this?

File: test/indexer/test_indexer.py [hunk 3]
```diff
@@ -351,7 +351,7 @@ def test_zero_tallies(self):
         for doc_id in doc_ids:
             message_re = re.compile(fr'INFO:elasticsearch:'
                                     fr'Got 201 response after [^ ]+ from PUT to '
-                                    fr'.*_aggregate/_doc/{doc_id}.*')
+                                    fr'.*_aggregate/_create/{doc_id}.*')
```
Comment (hannes-ucsc) [inline test/indexer/test_indexer.py]: This looks odd to me. `_doc` is what ES calls a `doctype`, the default one. They are deprecated, so maybe this is ok.

**Review by hannes-ucsc — COMMENTED** [review]
Replacing the boolean combinations with an enum makes sense. The code touched by that refactoring is rather complex and prone to breaking. I think we should do the refactoring as a separate PR and test it with a reindex in `prod`, without any other changes that would require a reindex.

File: src/azul/indexer/document.py [hunk 4]
```diff
@@ -1101,6 +1101,7 @@ class OpType(Enum):
 
     #: Remove the document from the index or fail if it does not exist
     delete = auto()
+    update = auto()
```
Comment (dsotirho-ucsc) [inline src/azul/indexer/document.py:1107]: `update` should get a description comment like the other OpTypes

File: src/azul/plugins/metadata/anvil/indexer/transform.py [hunk 5]
```diff
@@ -557,13 +561,14 @@ def _transform(self, entity: EntityReference) -> Transform:
             donors=self._entities(self._donor, linked['donor']),
             files=self._entities(self._file, linked['file']),
```
Comment (dsotirho-ucsc) [inline src/azul/plugins/metadata/anvil/indexer/transform.py]:
```suggestion
            files=self._entities(self._file, files),
```

**Review by dsotirho-ucsc — CHANGES_REQUESTED** [review]
Nicely done. Just a couple minor cleanup suggestions.

**Review by dsotirho-ucsc — DISMISSED** [review]
Approved.

**Comment by dsotirho-ucsc** [comment]
My force-push was to fix an accidental click of the `Update branch` button and remove the rebase commit added by GitHub.

**Review by dsotirho-ucsc — DISMISSED** [review]
Approved.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Conflicts.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
```diff
Subject: [PATCH] Add support for referencing images by digest
---
Index: src/azul/indexer/document.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/indexer/document.py b/src/azul/indexer/document.py
--- a/src/azul/indexer/document.py	(revision 91311a0e8c8f32c3db8ab29fd57138aae14e8173)
+++ b/src/azul/indexer/document.py	(date 1707327246928)
@@ -1103,8 +1103,8 @@
     delete = auto()
 
     #: Modify a document in the index via a scripted update or fail if it does
-    #: not exist.
-    update = auto()
+    #: not exist
+    scripted_update = auto()
 
 
 C = TypeVar('C', bound=DocumentCoordinates)
@@ -1311,7 +1311,8 @@
                 {}
                 if op_type is OpType.delete else
                 {
-                    '_source' if bulk else 'body': self._body(field_types[coordinates.entity.catalog])
+                    '_source' if bulk else 'body':
+                        self._body(field_types[coordinates.entity.catalog])
                 }
             ),
             '_id' if bulk else 'id': self.coordinates.document_id
@@ -1343,6 +1344,7 @@
         raise NotImplementedError
 
     def _body(self, field_types: FieldTypes) -> JSON:
+        # REVIEW: this should assert that optype is not update
         body = self.to_json()
         if self.needs_translation:
             body = self.translate_fields(doc=body,
@@ -1536,6 +1538,7 @@
         assert isinstance(self.coordinates, ReplicaCoordinates)
         assert self.coordinates.doc_type is DocumentType.replica
         assert self.entity.entity_type == 'replica', self.entity
+        # REVIEW: A comment should explain this
         self.hub_ids.sort()
 
     @classmethod
@@ -1570,6 +1573,8 @@
                     Collections.sort(ctx._source.hub_ids);
                 ''',
                     'params': {
+                        # REVIEW: Might be better to sort here but I'd still
+                        #         would like to know why sorting is necessary
                         'hub_ids': self.hub_ids
                     }
                 }
Index: src/azul/plugins/metadata/anvil/indexer/transform.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/metadata/anvil/indexer/transform.py b/src/azul/plugins/metadata/anvil/indexer/transform.py
--- a/src/azul/plugins/metadata/anvil/indexer/transform.py	(revision 91311a0e8c8f32c3db8ab29fd57138aae14e8173)
+++ b/src/azul/plugins/metadata/anvil/indexer/transform.py	(date 1707326964531)
@@ -488,7 +488,8 @@
             donors=self._entities(self._donor, self._entities_by_type['donor']),
             files=self._entities(self._file, files)
         )
-        return self._add_replica(contents, entity, [f.entity_id for f in files])
+        hub_ids = [f.entity_id for f in files]  # REVIEW: and elsewhere
+        return self._add_replica(contents, entity, hub_ids)
 
     @classmethod
     def field_types(cls) -> FieldTypes:
Index: src/azul/indexer/index_service.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/indexer/index_service.py b/src/azul/indexer/index_service.py
--- a/src/azul/indexer/index_service.py	(revision 91311a0e8c8f32c3db8ab29fd57138aae14e8173)
+++ b/src/azul/indexer/index_service.py	(date 1707326964540)
@@ -513,6 +513,11 @@
             retry_replicas = []
             for r in replicas:
                 if r.coordinates in writer.retries:
+                    # REVIEW: It feels like there is an assumption here about
+                    #         what the version type is. An assert could be added,
+                    #         but I think it would be better to rewrite the `if`
+                    #         to solely look at the version type first and then
+                    #         nest the `if` below
                     conflicts = writer.conflicts[r.coordinates]
                     if conflicts in (0, 1):
                         retry_replicas.append(r)
@@ -914,6 +919,9 @@
             log.debug('Successfully wrote %s with %i contribution(s).',
                       coordinates, doc.num_contributions)
         elif isinstance(doc, Replica) and doc.version_type is VersionType.none:
+            # REVIEW: Before this change, there would always be a log message.
+            # Now, there is a condition in which nothing is logged. If that is
+            # intentional, we should make that more clear (comments, asserts, …)
             log.debug('Successfully updated %s.', coordinates)
         else:
             log.debug('Successfully wrote %s.', coordinates)
```

**Comment by nadove-ucsc** [comment]
> \+    scripted_update = auto()

This will break because [the member name determines which method to call on the ES client](https://github.com/DataBiosphere/azul/blob/5f7406dd84eb9602e48dfc9f5aed72a1d200f1fe/src/azul/indexer/index_service.py#L856)

> \+            # REVIEW: Before this change, there would always be a log message.
\+            # Now, there is a condition in which nothing is logged. If that is
\+            # intentional, we should make that more clear (comments, asserts, …)

I don't believe that this is the case. What condition do you think will result in nothing being logged?

**Comment by nadove-ucsc** [comment]
> \+                        # REVIEW: Might be better to sort here but I'd still
\+                        #         would like to know why sorting is necessary

If hub ID's are only sorted during updates, then they will be unsorted when a replica is first written to the index and remain so until if/when a scripted update occurs. This breaks the unit tests by making the contents of the index non-deterministic.

**Comment by hannes-ucsc** [comment]
> This will break because [the member name determines which method to call on the ES client](https://github.com/DataBiosphere/azul/blob/5f7406dd84eb9602e48dfc9f5aed72a1d200f1fe/src/azul/indexer/index_service.py#L856)
> 

Makes sense. OK to leave it as is.

> I don't believe that this is the case. What condition do you think will result in nothing being logged?

None. I was mistaken.

File: src/azul/indexer/document.py [hunk 6]
```diff
@@ -1529,6 +1538,8 @@ def __attrs_post_init__(self):
         assert isinstance(self.coordinates, ReplicaCoordinates)
         assert self.coordinates.doc_type is DocumentType.replica
         assert self.entity.entity_type == 'replica', self.entity
+        # Ensure that index contents is deterministic for unit tests
+        self.hub_ids.sort()
```
Comment (hannes-ucsc) [inline src/azul/indexer/document.py]: Since `hub_ids` isn't immutable, please sort as late as possible, just before writing.

File: test/indexer/test_indexer.py [hunk 7]
```diff
@@ -480,9 +482,11 @@ def _assert_index_counts(self, *, just_deletion: bool):
         self._assert_hit_counts(hits,
                                 num_contribs=num_expected_addition_contributions + num_expected_deletion_contributions,
                                 num_aggs=num_expected_aggregates,
-                                # Indexing the same contribution twice (regardless of whether either is a deletion)
-                                # does not create a new replica, so we expect fewer replicas than contributions
-                                num_replicas=self._num_expected_replicas(num_expected_deletion_contributions))
+                                # Indexing the same contribution twice does not
```
Comment (hannes-ucsc) [inline test/indexer/test_indexer.py]: Please request PL slot to discuss this.

File: src/azul/plugins/metadata/hca/indexer/transform.py [hunk 8]
```diff
@@ -1693,7 +1694,7 @@ def _transform(self) -> Transform:
                         contributed_analyses=contributed_analyses,
                         dates=[self._date(self._singleton_entity())],
                         projects=[self._project(self._api_project)])
-        return self._add_replica(contents, self._singleton_entity())
+        return self._add_replica(contents, self._singleton_entity(), list(map(str, visitor.files)))
```
Comment (hannes-ucsc) [inline src/azul/plugins/metadata/hca/indexer/transform.py]: L 

Please extract variable as requested in previous review, for all calls to _add_replica

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/indexer/test_indexer.py (hunk 7 above)
Comment (nadove-ucsc) [reply test/indexer/test_indexer.py]: Feedback assimilated at PL.

File: src/azul/indexer/document.py [hunk 9]
```diff
@@ -1541,11 +1550,33 @@ def field_types(cls, field_types: FieldTypes) -> FieldTypes:
...
+        if self.op_type is OpType.update:
+            return {
+                'script': {
+                    'source': '''
+                    ctx._source.hub_ids.addAll(params.hub_ids);
+                    ctx._source.hub_ids = new ArrayList(new HashSet(ctx._source.hub_ids));
+                    Collections.sort(ctx._source.hub_ids);
+                ''',
+                    'params': {
+                        'hub_ids': sorted(self.hub_ids)
```
Comment (hannes-ucsc) [inline src/azul/indexer/document.py]: Why do you sort here? I don't think `params` manifests anywhere in test code or cans.

File: src/azul/indexer/document.py [hunk 10]
```diff
@@ -1541,11 +1550,33 @@ def field_types(cls, field_types: FieldTypes) -> FieldTypes:
     def to_json(self) -> JSON:
         return dict(super().to_json(),
                     replica_type=self.replica_type,
-                    hub_ids=self.hub_ids)
+                    hub_ids=sorted(self.hub_ids))
```
Comment (hannes-ucsc) [inline src/azul/indexer/document.py]: A comment should explain why, just like in the previous version.

File: src/azul/indexer/document.py [hunk 11]
```diff
@@ -1541,11 +1550,33 @@ def field_types(cls, field_types: FieldTypes) -> FieldTypes:
...
+            return OpType.update
+        else:
+            assert False, self.version_type
+
+    def _body(self, field_types: FieldTypes) -> JSON:
+        if self.op_type is OpType.update:
+            return {
+                'script': {
+                    'source': '''
+                    ctx._source.hub_ids.addAll(params.hub_ids);
```
Comment (hannes-ucsc) [inline src/azul/indexer/document.py]:
```suggestion
                    Stream stream = Stream.concat(ctx._source.hub_ids.stream(), params.hub_ids.stream());
                    ctx._source.hub_ids = stream.sorted().distinct().collect(Collectors.toList());
```

https://stackoverflow.com/questions/43806467/java-streams-how-to-do-an-efficient-distinct-and-sort

File: test/indexer/test_indexer.py [hunk 12]
```diff
@@ -2113,6 +2119,31 @@ def test_disallow_manifest_column_joiner(self):
...
+                                                                          entity_type='replica',
+                                                                          entity_id='foo'))
+        replica = Replica(version=None,
+                          replica_type='file',
+                          contents=contents,
+                          hub_ids=[],
+                          coordinates=coordinates)
+
+        for case, hub_ids, expected_hub_ids in [
+            ('New replica', ['1'], ['1']),
```
Comment (hannes-ucsc) [inline test/indexer/test_indexer.py]: Should also test duplicate hub ids in new replicas.

File: test/indexer/test_indexer.py [hunk 13]
```diff
@@ -2113,6 +2119,31 @@ def test_disallow_manifest_column_joiner(self):
...
+            ('Additional hub IDs', ['1', '2', '3'], ['1', '2', '3']),
```
Comment (hannes-ucsc) [inline test/indexer/test_indexer.py]: Reverse the order of the hub_ids to make this stronger.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]
